Fix out-of-bound access in ip6addr_ntoa_r()

sfionov opened the pull request:
When detecting that zero is single, code reads the next group even if current group is last group.

If next bytes are non-null, last zero is not omitted.

If next bytes are null, last zero is omitted, but since there are no groups left, finishing ':' will not be written, resulting in invalid address.

This commit turns off non-single zero check for the last group.

sfionov commented on src/core/ipv6/ip6_addr.c:
```diff
@@ -270,15 +270,16 @@ ip6addr_ntoa_r(const ip6_addr_t *addr, char *buf, int buflen)
+      } else if (empty_block_flag == 0) {
         /* generate empty block "::", but only if more than one contiguous zero block,
          * according to current formatting suggestions RFC 5952. */
         next_block_value = lwip_htonl(addr->addr[(current_block_index + 1) >> 1]);
```
Out-of-bounds access is happened here.

yarrick commented:
The test passes even without the ip6_addr code change. Could you split it into a commit with a failing test, then a commit which fixes the code?

yarrick commented:
The updated test is still passing when skipping the ip6_addr code change. Can you split into a commit with a failing test and then the change that fixes it?

sfionov committed:
Fix out-of-bound access in ip6addr_ntoa_r()

When detecting that zero is single, code reads the next group even if current group is last group.

If next bytes are not-null, last zero is not omitted.

If next bytes are null, last zero is omitted, but since there are no groups left,
finishing ':' will not be written, resulting in invalid address.

This commit turns off non-single zero check for the last group.

sfionov committed:
Add test for out-of-bound access in ip6addr_ntoa_r()

sfionov commented:
@yarrick Done, split into two commits

yarrick commented:
The test still passes when applied on top of the current code

sfionov commented:
@yarrick What platform do you use? 

I originally detected this issue when arm64 musl-gcc 13.2 refused to compile code with -Wall -Wextra. 

And test fails on arm64 macOS without patch.

yarrick commented:
I run on Linux on x86_64 (Fedora 40)
